src/engine/GameObjects/MapNode.cxx

```diff
@@ -49,25 +49,53 @@ void MapNode::setBitmask(unsigned char elevationBitmask, unsigned char tileIDBit
 
 void MapNode::setTileID(const std::string &tileID)
 {
-  if (TileManager::instance().getTileData(tileID))
+  TileData *tileData = TileManager::instance().getTileData(tileID);
+  if (tileData)
   {
-    if (TileManager::instance().getTileData(tileID)->category == "Terrain" ||
-        TileManager::instance().getTileData(tileID)->category == "Water")
+    Layer layer = Layer::BUILDINGS;
+    if (tileData->category == "Terrain")
     {
-      m_previousTileID = m_mapNodeData[Layer::TERRAIN].tileID;
-      m_mapNodeData[Layer::TERRAIN].tileData = TileManager::instance().getTileData(tileID);
-      m_mapNodeData[Layer::TERRAIN].tileID = tileID;
+      layer = Layer::TERRAIN;
     }
-    else
+    else if (tileData->category == "Water")
     {
-      m_previousTileID = m_mapNodeData[Layer::BUILDINGS].tileID;
-      m_mapNodeData[Layer::BUILDINGS].tileData = TileManager::instance().getTileData(tileID);
-      m_mapNodeData[Layer::BUILDINGS].tileID = tileID;
+      layer = Layer::WATER;
     }
+    m_previousTileID = m_mapNodeData[layer].tileID;
+    m_mapNodeData[layer].tileData = tileData;
+    m_mapNodeData[layer].tileID = tileID;
     updateTexture();
   }
 }
 
+bool MapNode::checkTileIsEmpty(const std::string &tileID) const
+{
+  TileData *tileData = TileManager::instance().getTileData(tileID);
+  if (tileData)
+  {
+    Layer layer = Layer::BUILDINGS;
+    if (tileData->category == "Terrain")
```

Comment:
I think this should be done in serialization...

Reply:
In serialization, we can use a trie instead of a hashmap. Tries are very good for string matching. Otherwise a small finite state machine is equivalent

Reply:
Forget what I said about tries. [Use better enums](https://github.com/nlohmann/json#specializing-enum-conversion)